ocaml/xapi/xapi_vdi.ml

```diff
@@ -758,4 +758,84 @@ let read_database_pool_uuid ~__context ~self =
 	| Some (_, uuid) -> uuid
 	| None -> ""
 
+module VDI_CStruct = struct
+
+	let magic_number = 0x7ada7adal
+	let magic_number_offset = 0
+	let version = 1l
+	let version_offset = 4
+	let length_offset = 8
+	let data_offset = 12
+	let vdi_format_length = 12 (* VDI format takes 12bytes *)
+	let vdi_size = 4194304 (* 4MiB *)
+	let default_offset = 0
+
+	(* Set the magic number *)
+	let set_magic_number cstruct =
+		Cstruct.BE.set_uint32 cstruct magic_number_offset magic_number
+
+	(* Get the magic number *)
+	let get_magic_number cstruct =
+		Cstruct.BE.get_uint32 cstruct magic_number_offset
+
+	(* Set the version *)
+	let set_version cstruct =
+		Cstruct.BE.set_uint32 cstruct version_offset version
+
+	(* Set the data length *)
+	let set_data_length cstruct len =
+		Cstruct.BE.set_uint32 cstruct length_offset len
+
+	(* Get the data length *)
+	let get_data_length cstruct =
+		Cstruct.BE.get_uint32 cstruct length_offset
+
+	(* Write the string to the vdi *)
+	let write_to_vdi cstruct text text_len =
+		Cstruct.blit_from_string text default_offset cstruct data_offset text_len;
+		set_data_length cstruct (Int32.of_int text_len)
+
+	(* Read the string from the vdi *)
+	let read_from_vdi cstruct =
+		let curr_len = Int32.to_int (get_data_length cstruct) in
+		let curr_text = String.make curr_len '\000' in
+		Cstruct.blit_to_string cstruct data_offset curr_text default_offset curr_len;
+		curr_text
+
+	(* Format the vdi for the first time *)
+	let format_rrd_vdi cstruct =
+		set_magic_number cstruct;
+		set_version cstruct
+
+end
+
+(* Write the rrd stats to the rrd-stats vdi *)
+let write_rrd ~__context ~sr ~vdi ~text =
+	Helpers.call_api_functions ~__context
+		(fun rpc session_id -> Sm_fs_ops.with_open_block_attached_device __context rpc session_id vdi `RW
+			(fun fd ->
+				let mapping = Bigarray.(Array1.map_file fd char c_layout true VDI_CStruct.vdi_size) in
+				let cstruct = Cstruct.of_bigarray mapping in
+				if (VDI_CStruct.get_magic_number cstruct) <> VDI_CStruct.magic_number then
+					VDI_CStruct.format_rrd_vdi cstruct;
+				let new_text_len = String.length text in
+				if new_text_len >= (VDI_CStruct.vdi_size - VDI_CStruct.vdi_format_length) then
+					raise (Api_errors.Server_error(Api_errors.vdi_out_of_space, [ Ref.string_of vdi ]));
+				VDI_CStruct.write_to_vdi cstruct text new_text_len
+			)
+		)
+
+(* Read the rrd stats from the rrd-stats vdi *)
+let read_rrd ~__context ~sr ~vdi =
+	Helpers.call_api_functions ~__context
+		(fun rpc session_id -> Sm_fs_ops.with_open_block_attached_device  __context rpc session_id vdi `RW
```

Comment:
Since we're only going to be reading from this block device, it can be opened `` `RO ``.